src/ray/raylet/local_object_manager.h

```diff
@@ -0,0 +1,126 @@
+// limitations under the License.
+
+#pragma once
+
+#include <google/protobuf/repeated_field.h>
+
+#include <functional>
+
+#include "ray/common/id.h"
+#include "ray/common/ray_object.h"
+#include "ray/gcs/accessor.h"
+#include "ray/raylet/worker_pool.h"
+#include "ray/rpc/worker/core_worker_client_pool.h"
+
+namespace ray {
+
+namespace raylet {
+
+/// This class implements memory management for primary objects, objects that
+/// have been freed, and objects that have been spilled.
+class LocalObjectManager {
+ public:
+  LocalObjectManager(size_t free_objects_batch_size, int64_t free_objects_period_ms,
+                     IOWorkerPoolInterface &io_worker_pool,
+                     gcs::ObjectInfoAccessor &object_info_accessor,
+                     rpc::CoreWorkerClientPool &owner_client_pool,
+                     std::function<void(const std::vector<ObjectID> &)> on_objects_freed)
+      : free_objects_period_ms_(free_objects_period_ms),
+        free_objects_batch_size_(free_objects_batch_size),
+        io_worker_pool_(io_worker_pool),
+        object_info_accessor_(object_info_accessor),
+        owner_client_pool_(owner_client_pool),
+        on_objects_freed_(on_objects_freed),
+        last_free_objects_at_ms_(current_time_ms()) {}
+
+  /// Pin objects.
+  ///
+  /// The objects will be released when the owner at the given address fails or
+  /// replies that the object has been evicted.
+  ///
+  /// \param owner_address The address of the owner of the objects.
+  /// \param object_ids The objects to be pinned.
+  /// \param objects Pointers to the objects to be pinned. The pointer should
+  /// be kept in scope until the object can be released.
+  void PinObjects(const rpc::Address &owner_address,
+                  const std::vector<ObjectID> &object_ids,
+                  std::vector<std::unique_ptr<RayObject>> &&objects);
+
+  /// Spill objects to external storage.
+  ///
+  /// \param objects_ids_to_spill The objects to be spilled.
+  /// \param callback A callback to call once the objects have been spilled, or
+  /// there is anerror.
+  void SpillObjects(const std::vector<ObjectID> &objects_ids,
+                    std::function<void(const ray::Status &)> callback);
+
+  /// Restore a spilled object from external storage back into local memory.
+  ///
+  /// \param object_id The ID of the object to restore.
+  /// \param object_url The URL in external storage from which the object can be restored.
+  /// \param callback A callback to call when the restoration is done. Status
+  /// will contain the error during restoration, if any.
+  void AsyncRestoreSpilledObject(const ObjectID &object_id, const std::string &object_url,
+                                 std::function<void(const ray::Status &)> callback);
+
+  /// Try to delete any objects that have been freed.
+  void FlushIfNeeded(int64_t now_ms);
+
+ private:
+  /// Delete freed objects.
+  void Flush();
```

Comment:
Are we planning to spill those freed objects later within this method? The function name is a little confusing to me (because as far as I know flush means you'd like to move buffered objects onto the durable storage).

Reply:
This actually just triggers the callback for freed objects. It doesn't have anything to do with spilling. I'll update the name and description.